sdk/core/Azure.Core/src/Pipeline/RetryPolicy.cs

```diff
@@ -0,0 +1,331 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Diagnostics;
+
+namespace Azure.Core.Pipeline
+{
+    /// <summary>
+    /// Represents a policy that can be overriden to customize whether or not a request will be retried and how long to wait before retrying.
+    /// </summary>
+    public abstract class RetryPolicy : HttpPipelinePolicy
+    {
+        private readonly RetryMode _mode;
```

Comment:
I don't think this would work since the user may not override this method, so we would still need the logic to live in the base public class and have it respect the options. But perhaps I'm misunderstanding.

Reply:
I like the idea of making internal subclasses do something, and I'd be interested to see how much you could simplify the implementation by doing this.

If we don't do that, do we even need to make this abstract with an internal implementation?  People could just subclass the type and optionally override virtual methods.

Reply:
Fine with making this non-abstract if that is what the consensus is. I preferred abstract because IMO it communicates to users the intent of the type - it is meant to be overriden. 

As for using the two subclasses for exponential and fixed, as I mentioned above, I don't think this will work given that customers can subclass the policy without implementing the CalculateNextDelay method.

Reply:
> I don't think this will work given that customers can subclass the policy without implementing the CalculateNextDelay method.

Hm, that is kind of interesting.  Have we clarified and documented the expected semantics regarding how a replaced RetryPolicy would work with any RetryOptions set on ClientOptions?

Reply:
I will document this. The way it works is that the options can be passed into the policy so the user can choose to respect the options for any overridden methods. For methods that are not overridden, the base policy will respect the options.

Reply:
Added a sample - https://github.com/Azure/azure-sdk-for-net/pull/31187/files#diff-72428161e5066dbbc5e348e6031475d84f40c2b1241d6d4d8503e945bf217469R25